draft-ietf-quic-recovery.md

```diff
@@ -1407,25 +1407,26 @@ Invoked when an ACK frame with an ECN section is received from the peer.
 Invoked from DetectLostPackets when packets are deemed lost.
 
 ~~~
-   InPersistentCongestion(largest_lost_packet):
+   InPersistentCongestion(earliest_lost_packet, largest_lost_packet):
      pto = smoothed_rtt + max(4 * rttvar, kGranularity) +
        max_ack_delay
      congestion_period = pto * kPersistentCongestionThreshold
      // Determine if all packets in the time period before the
      // newest lost packet, including the edges, are marked
      // lost
-     return AreAllPacketsLost(largest_lost_packet,
-                              congestion_period)
+     return congestion_period >
+       largest_lost_packet.time_sent - earliest_lost_packet.time_sent
```

Comment:
I think it's possible a packet is acked between earliest and largest, so I believe you may need to pass in lost_packets and check that they're contiguous.  Then you can do an early return if no packets are lost or if they're not contiguous.

ie: 
all_packets_lost =  acket_packets.length() ==
   (acked_packets.last().packet_number - acked_packets.first().packet_number - 1)

Reply:
Done I think, though did `s/acked_packets/lost_packets/` (also got a bit creative with the indentation).

Reply:
Sorry, I meant lost_packets.

Reply:
Also, is the `- 1` really needed here?

Reply:
+1 is.  

Think through the example where 0, 1 and 2 are lost.  lost_packets has length 3.  But 2 - 0 = 2, so you need to add 1.

Reply:
But also, now that I think about it, `lost_packets` only contains lost in-flight packets right? So checking `lost_packets.length()` against the packet number difference might fail because of other lost but not in-flight packets adding gaps, is this a problem?

Reply:
Good point, that is a problem.  I think that was why we resorted to a helper method to begin with.

For a packet between smallest and largest to be acknowledged AND smallest not previously declared lost, I believe it must occur in the ACK frame that caused smallest and largest to be lost.  Otherwise the time threshold would have kicked in long before persistent congestion.

Therefore, if the largest lost packet is smaller than the earliest newly acknowledged packet, we can guarantee the lost packets are contiguous.  However, then we do not trigger persistent congestion if a packet sent at the very beginning of the congestion period is acknowledged in a long delayed ACK, and at the same time a much later packet is ACKed.  I'm not sure we care about capturing that case, but it's worth considering.

If we were ok with more complex code, we could loop over all newly acked packets and ensure none were between the smallest and largest lost packets.

Reply:
Isn't it `congestion_period <= largest_lost_packet.time_sent - earliest_lost_packet.time_sent` because congestion_period is the threshold that defines duration of persistent congestion?